xla/service/gpu/loop_double_buffer_transformer.cc

```diff
@@ -0,0 +1,372 @@
+            TF_RETURN_IF_ERROR(new_pred->AddControlDependencyTo(new_instr));
+            VLOG(2) << "Adding " << new_pred->ToString()
+                    << " to control dependency of " << new_instr->ToString();
+          }
+        }
+      }
+    }
+    for (auto instr : all_new_instructions) {
+      for (int64_t op_index = 0; op_index < instr->operand_count();
+           op_index++) {
+        VLOG(2) << "Processing operand " << op_index << " of "
+                << instr->ToString();
+        HloInstruction* operand = instr->mutable_operand(op_index);
+        if (operand == input_parameter || operand->IsUserOf(input_parameter)) {
+          VLOG(2) << "Operand " << op_index << " is an input parameter.";
+          continue;
+        }
+        VLOG(2) << "Changing operand " << op_index << " of instruction "
+                << instr->ToString() << " from " << operand->ToString()
+                << " to " << old_to_new_map[operand]->ToString();
+        TF_RETURN_IF_ERROR(
+            instr->ReplaceOperandWith(op_index, old_to_new_map[operand]));
+      }
+    }
+
+    // Connect the boundary, meaning adding previous roots as operands of
+    // instructions in the next iteration
+    for (auto& instr_index_pair : loop_input_consumers) {
+      VLOG(2) << "Processing boundary instruction: "
+              << instr_index_pair.first->ToString();
+      HloInstruction* instr = old_to_new_map[instr_index_pair.first];
+      // Get the operand index of its consumer
+      for (auto consumer_info : instr_index_pair.second) {
+        int64_t op_index = consumer_info.operand_index;
+        HloInstruction* new_operand = old_loop_roots[consumer_info.tuple_index];
+        VLOG(2) << "Changing boundary operand " << op_index
+                << " of instruction " << instr->ToString() << " from "
+                << instr->operand(op_index)->ToString() << " to "
+                << new_operand->ToString();
+        if (instr->operand(op_index) != new_operand) {
+          TF_RETURN_IF_ERROR(instr->ReplaceOperandWith(op_index, new_operand));
+        }
+      }
+      if (skip_control_dep_injection.find(instr_index_pair.first) ==
+              skip_control_dep_injection.end() &&
+          !IsCollective(instr)) {
+        for (auto& old_root : old_loop_roots) {
+          TF_RETURN_IF_ERROR(old_root->AddControlDependencyTo(instr));
+        }
+      }
+    }
+
+    // Make a new tuple as the new root
+    for (int64_t i = 0; i < loop_output_count; i++) {
+      if (old_to_new_map.find(old_loop_roots[i]) != old_to_new_map.end()) {
+        new_loop_roots[i] = old_to_new_map[old_loop_roots[i]];
+      } else {
+        new_loop_roots[i] = old_loop_roots[i];
+      }
+    }
+    TF_RETURN_IF_ERROR(while_body->ReplaceWithNewInstruction(
+        while_body->root_instruction(),
+        HloInstruction::CreateTuple(new_loop_roots)));
+
+    WhileLoopBackendConfig new_config;
+    new_config.mutable_known_trip_count()->set_n((exact_trip_count / 2));
```

Comment:
This seems to say that calling the `WhileLoopTripCountAnnotator` after this pass is not necessary, correct?

Reply:
right, this is only thing that changes so no need to run WhileLoopTripCountAnnotator anymore